main.py

```diff
@@ -345,31 +362,99 @@ def writer(args, res_queue, input_file_path):
-                    open(os.path.join(out_base_path + ".peaks.bedGraph"), "w")]
+        outfiles = [os.path.join(out_base_path + ".track.bedGraph"),
+                    os.path.join(out_base_path + ".peaks.bedGraph")]
+        rounding = [3, 3]
+
+    # Temp dir used to save temp files during multiprocessing.
+    temp_dir = tempfile.mkdtemp()
+    for channel in channels:
+        os.makedirs(os.path.join(temp_dir, str(channel)))
+
+    batch_num = 0
     while 1:
         if not res_queue.empty():
+            batch_num = batch_num + 1
             item = res_queue.get()
             if item == 'done':
                 break
-            for channel in channels:
-                batch_bg = get_batch_bg(args, item, channel, intervals)
-                df_to_bedGraph(batch_bg, outfiles[channel])
+            keys, batch = item
+            #No multi processing
+            if args.num_workers == 0:
+                start = 0
+                end = batch.shape[0]
+                for channel in channels:
+                    batch_bg = save_to_bedgraph([start, end], item, channel, intervals, outfiles[channel], args.round)
+            else:
+                num_jobs = math.ceil(batch.shape[0] / args.batches_per_worker)
+                pool_size = 0
+
+                if args.num_workers == -1: # spawn pool of processes
+                    num_cpus = mp.cpu_count()
+                    pool_size = min(num_jobs, num_cpus)
+                else: # user specified # of processes
+                    pool_size = args.num_workers
+
+                pool = mp.Pool(pool_size)
+                tmp_batch_ranges = [[i*args.batches_per_worker, (i+1)*args.batches_per_worker] for i in range(num_jobs)]
+                # Force the last interval to capture remaining data.
+                tmp_batch_ranges[-1][1] = batch.shape[0]
+                all_intervals = [intervals]*len(tmp_batch_ranges)
+                all_items = [item]*len(tmp_batch_ranges)
+                for channel in channels:
+                    temp_files = [os.path.join(temp_dir, str(channel), "{0:05}".format(num+batch_num)) for num in range(num_jobs)]
+                    if args.infer_threshold is None:
+                        map_args = list(zip(tmp_batch_ranges, all_items,
+                                            [channel]*len(tmp_batch_ranges), all_intervals,
+                                            temp_files, [rounding[channel]]*len(tmp_batch_ranges)))
+                    else:
+                        map_args = list(zip(tmp_batch_ranges, all_items,
+                                            [channel]*len(tmp_batch_ranges), all_intervals,
+                                            temp_files, [rounding[channel]]*len(tmp_batch_ranges),
+                                            [args.infer_threshold]*len(tmp_batch_ranges)))
+
+                    pool.starmap(save_to_bedgraph, map_args)
+
+                for channel in channels:
```

Comment:
Just noticed another thing - we can put this combiner within the same for loop above, so we wrote and combine for a single channel at a time

Reply:
done.